nf_core/download.py

```diff
@@ -0,0 +1,197 @@
+        if len(wf.containers) == 0:
+            logging.info("No container names found in workflow")
+        else:
+            os.mkdir(os.path.join(wf.outdir, 'singularity-images'))
+            for container in wf.containers:
+                wf.download_singularity_image(container)
+
+class DownloadWorkflow():
+
+    def __init__(self, pipeline, release=None, singularity=False, outdir=None):
+        """ Set class variables """
+
+        self.pipeline = pipeline
+        self.release = release
+        self.singularity = singularity
+        self.outdir = outdir
+
+        self.wf_name = None
+        self.wf_sha = None
+        self.wf_download_url = None
+        self.containers = list()
+
+
+    def get_workflow(self):
+        """ Fetch details of nf-core workflow to download """
+        wfs = nf_core.list.Workflows()
+        wfs.get_remote_workflows()
+
+        # Get workflow download details
+        for wf in wfs.remote_workflows:
+            if wf.full_name == self.pipeline or wf.name == self.pipeline:
+
+                # Set pipeline name
+                self.wf_name = wf.name
+
+                # Find latest release hash
+                if self.release is None and len(wf.releases) > 0:
+                    self.release = wf.releases[0]['tag_name']
+                    self.wf_sha = wf.releases[0]['tag_sha']
+                    logging.debug("No release specified. Using latest release: {}".format(self.release))
+                # Find specified release hash
+                elif self.release is not None:
+                    for r in wf.releases:
+                        if r['tag_name'] == self.release.lstrip('v'):
+                            self.wf_sha = r['tag_sha']
+                            break
+                    else:
+                        logging.error("Not able to find release '{}' for {} (found '{}')".format(self.release, wf.full_name, "', '".join([r['tag_name'] for r in wf.releases])))
+                        return False
+                # Must be a dev-only pipeline
+                elif self.release is None:
+                    self.release = 'dev'
+                    self.wf_sha = 'master' # Cheating a little, but GitHub download link works
+                    logging.info("Pipeline is in development. Using current code on master branch.")
+
+                # Set outdir name if not defined
+                if self.outdir is None:
+                    self.outdir = 'nf-core-{}'.format(wf.name)
+                    if self.release is not None:
+                        self.outdir += '-{}'.format(self.release)
+
+                # Set the download URL and return
+                self.wf_download_url = 'https://github.com/{}/archive/{}.zip'.format(wf.full_name, self.wf_sha)
+                return True
+
+        # If we got this far, must not be a nf-core pipeline
+        if self.pipeline.count('/') == 1:
+            # Looks like a GitHub address - try working with this repo
```

Comment:
Why not check this upstream ? i.e. validate the param value at very beginning. That could improve the success rate of downloading something even higher 😉

Also could be good to log that it is not a `nf-core` pipeline (?)

Reply:
Yup, should do both 👍 Was rushing to push the code so that I wasn't late into work for Remi's talk 😉

Reply:
This is as early as I can check I think. In order to validate the pipeline name, I have to fetch the pipeline names from the website API and see if the supplied name matches any of them (the above `for wf in wfs.remote_workflows` loop). So this point in the code is the first time that I know it's _not_  an nf-core pipeline.

Agree that logging is a good idea 👍

Reply:
Logging added in fd90e46ef55c585871c08b9cb0119fbea2e0d7b3